### ISSUE TITLE
[RFC] Don't pass attr, obj, data to fields (de)serialize methods

### ISSUE THREAD

**Issue by lafrech** [marshmallow-code/marshmallow#2039]
From a quick look, those are not used by existing fields. I suspect this might be or have been useful in specific fields such as `Method`.

I can't investigate this thoroughly right now and anyway that would be a breaking change, but opening this here for discussion.

For separation of concerns, (de)ser methods shouldn't need the parent obj/data and field name.

Also, this means that one needs an object and a field name to serialize some data with a field, which may not always be the case. They may pass `None` but then they don't respect the API (since it won't work with a field (de)ser method that actually relies on those). I faced this specific problem with `EnumValue` field, when building `self.choices`, see #2017.

**Comment by lafrech** [comment]
I gave it a quick look.

`attr` and `obj` are needed in `serialize` to get the value from the object.

In `_serialize`, `deserialize` and `_deserialize`, they seem to be only used by `Function` / `Method` fields, so removing those would allow us to simplify the implementation.

I guess some users rely on those fields. OTOH, if they didn't exist, I doubt we'd modify (de)ser methods to pass `attr` and `obj` just to add them.

I don't have a simple idea to replace those fields, though. pre_load/dump methods don't really answer because we don't want to modify the object. Unless we add some sort of proxy around the object but this sounds complicated.

**Comment by lafrech** [comment]
An option could be to remove `Function` and `Method` fields and add an argument or two to `Field` to allow the user to pass a custom function/method to get the value from the object. This function/method could be called from `get_value`. In fact, it would be a `get_value` override.

So it wouldn't be a field by itself but rather a customization of an existing field. There would be a benefit in terms of auto-documentation, since the type of the value would be known (the user may choose to use `Field` for a type-free value).

We could add this feature and deprecate `Function` and `Method` (as well as the use of `attr` and `obj`) in marshmallow 3.x.

-----------------------------

I was thinking of opening another issue about moving `get_value` logic to the schema. The schema would just need to access the `attribute` attribute of the field. If we do what I describe above, the schema would also need to call the custom accessor (the function/method). It's not obvious to me where the logic should be, but I still find it strange to pass an obj, an accessor function, etc.

**Comment by lafrech** [comment]
We could add `accessor` parameter to `Field`.

```py
class Field(FieldABC):
   def __init__(..., accessor,...):
        ...
        self.get_value = accessor
        ...
```

Allowing the user to do that:

```py

def ser_function(...):
    ...

class MySchema(Schema):

    def ser_method(...):
        ...

    # Deprecated
    old_function = fields.Function(ser_function)
    # Replace with
    new_function = fields.Field(accessor=ser_function)

    # Deprecated
    old_method = fields.Method(ser_method)
    # Replace with
    new_method = fields.Field(accessor=ser_method)
```

**Comment by lafrech** [comment]
I still need to get my head around this. If `Function`/`Method`/`Constant` used `serialize` rather than `_serialize`, we could remove the `_CHECK_ATTRIBUTE` hack.

But then we wouldn't call `get_value` so the simple solution above wouldn't work.

Also, it is worth noting that fields working at `get_value` level couldn't be embedded in container fields, as those are meant to pass the value, not the object (according to this issue and https://github.com/marshmallow-code/marshmallow/issues/1995#issuecomment-1281451194). Perhaps we can live with this limitation. I don't see the point of using such a field in a container field.

**Comment by lafrech** [comment]
`Function` / `Method` / `Constant` address two use cases.

- 1/ Define custom fields as one-liners. This is by no means necessary but can arguably be desired.
- 2/ Define fields that access the value from the object in a fancy way. This is what requires passing `attr` and `obj`.

We could keep `Function` and `Method` for case 1 without passing `attr` and `obj`. Or even decide that defining a custom field the normal way is acceptable and drop them altogether.

Case 2 could be addressed by offering a way to pass custom accessors to any field constructor. We could even offer to define one for loading and one for dumping. As I wrote above, the user would benefit from the field being correctly typed and offering correct validation, etc.

**This would separate access and (de)serialization concerns.**

So basically,

- `Function` and `Method` would keep their ser/deser functions/methods but those would receive the value after it has been accessed and wouldn't access the object. This way, they would work correctly in container fields as well.

- `Field` would receive a `get_value` parameter to override the way the value is accessed from the object. Maybe even one for deserialization if we want to go further than what is possible today. `serialization_getter` and `deserialization_getter`.

- `Constant` could be addressed by any field of the right type with a custom accessor.

I think this may finally be the way to get this right. Thoughts welcome.

**Comment by panda-byte** [comment]
I think I have an issue in another way, because I would really like to be able to access the actual passed value, but this is not possible now, right? 

```python
class MySchema(Schema):
    def serialize(value, attr, obj): str  # ❌ not possible
        return f"field {attr} has value {value}

    field_1 = Method(serialize)
    field_2 = Method(serialize)
```

In the above use case I would like to apply the same method on multiple fields of the same kind. As it stands right now, I would need to create a custom field for that, right? Why not pass all three arguments from `_serialize`? Because the ser/deser methods need to be defined directly on the parent anyways, I would assume developers would know exactly what kind of arguments to expect, because the methods are meant to be used for a precise set of circumstances. Even if developers would use `Method` for some container field or something, would it be a problem at all if e.g. `attr` was `None`? I can't really see the problem here, I think I'm approaching this from a completely different direction. And if you don't want your function to take all the arguments, you could easily bypass this with `field_x = Method(lambda _, _, obj: my_method_only_likes_obj(obj)` or similar

**Comment by lafrech** [comment]
I think the change I'm proposing does what you want (pass value to serialization method).
See #2963. But don't hold your breath. This is a pretty big breaking change.

**Comment by lafrech** [comment]
Once dump/load_getter provide fancy accessing, and since fields now have pre/post-dump/load processors (#2787), `Function` and `Method` shouldn't actually be useful anymore.
